File: matlab_code/Denoising.m
function re = Denoising(resource, waveFrequency)
    %%去除基线漂移
    [C,L] = wavedec(resource,8,'sym8');%8层小波分解
    A8=C(1:L(1)); 
    D8=C(L(1)+1:L(1)+L(2));%去掉注释后可以得到论文中图3-5
    D7=C(L(1)+L(2)+1:L(1)+L(2)+L(3));
    D6=C(L(1)+L(2)+L(3)+1:L(1)+L(2)+L(3)+L(4));
    D5=C(L(1)+L(2)+L(3)+L(4)+1:L(1)+L(2)+L(3)+L(4)+L(5));
    D4=C(L(1)+L(2)+L(3)+L(4)+L(5)+1:L(1)+L(2)+L(3)+L(4)+L(5)+L(6));
    D3=C(L(1)+L(2)+L(3)+L(4)+L(5)+L(6)+1:L(1)+L(2)+L(3)+L(4)+L(5)+L(6)+L(7));
    D2=C(L(1)+L(2)+L(3)+L(4)+L(5)+L(6)+L(7)+1:L(1)+L(2)+L(3)+L(4)+L(5)+L(6)+L(7)+L(8));
    D1=C(L(1)+L(2)+L(3)+L(4)+L(5)+L(6)+L(7)+L(8)+1:L(1)+L(2)+L(3)+L(4)+L(5)+L(6)+L(7)+L(8)+L(9));
    CC=zeros(size(C));
    CC(1:L(1))=A8;

    baseline = waverec(CC,L,'sym8');%去除细节因子Dl一Dg后得到的基线漂移
    x1 = resource - baseline;%去除基线飘移后的信号

    %%butter带通滤波
    Wp=[0.9 50]/500;Ws=[0.3 140]/500;
    [n,Wn] = buttord(Wp,Ws,3,10);
    [b,a] = butter(n,Wn);
    %freqz(b,a,512,waveFrequency);
    
    %%cheby去除50hz工频干扰
    Wp1=[30 65]/500;Ws1=[45 55]/500;
    rp=3;rs=60;
    [n1,Wn1] = cheb2ord(Wp1,Ws1,rp,rs);
    [b1,a1] = cheby2(n1,rs,Wn1,'stop');
    %freqz(b1,a1,512,waveFrequency);
    x2=filtfilt(b,a,x1);
    re=filtfilt(b1,a1,x2);

File: python/Denoising.py
import pywt
import numpy as np
import scipy.signal as sci
import matplotlib.pyplot as plt

def denoising(resource, NOISE_LEVEL, WAVELET_LEVEL):
    try:
        C = pywt.wavedec(resource, 'sym8',level=WAVELET_LEVEL);

        # If the noise is not very large, then comment the following two lines
        if NOISE_LEVEL == 2:
            C[5] = np.zeros(C[5].size);
            C[6] = np.zeros(C[6].size);
            C[7] = np.zeros(C[7].size);
            C[8] = np.zeros(C[8].size);
        elif NOISE_LEVEL == 1:
            C[7] = np.zeros(C[7].size);
            C[8] = np.zeros(C[8].size);

        denoiseWave = pywt.waverec(C,'sym8');
        C2 = pywt.wavedec(denoiseWave, 'sym8',level=WAVELET_LEVEL);
        C2[1] = np.zeros(C2[1].size);
        C2[2] = np.zeros(C2[2].size);
        C2[3] = np.zeros(C2[3].size);
        C2[4] = np.zeros(C2[4].size);
        C2[5] = np.zeros(C2[5].size);
        C2[6] = np.zeros(C2[6].size);
        C2[7] = np.zeros(C2[7].size);
        C2[8] = np.zeros(C2[8].size);
        baseline = pywt.waverec(C2,'sym8');
        x1 = denoiseWave - baseline;

    except ValueError:
        C = pywt.wavedec(resource, 'sym8',level=WAVELET_LEVEL-1);
        # If the noise is not very large, then comment the following two lines
        if NOISE_LEVEL == 2:
            C[4] = np.zeros(C[4].size);
            C[5] = np.zeros(C[5].size);
            C[6] = np.zeros(C[6].size);
            C[7] = np.zeros(C[7].size);
        elif NOISE_LEVEL == 1:
            C[6] = np.zeros(C[6].size);
            C[7] = np.zeros(C[7].size);

        denoiseWave = pywt.waverec(C,'sym8');
        C2 = pywt.wavedec(denoiseWave, 'sym8',level=WAVELET_LEVEL-1);
        C2[1] = np.zeros(C2[1].size);
        C2[2] = np.zeros(C2[2].size);
        C2[3] = np.zeros(C2[3].size);
        C2[4] = np.zeros(C2[4].size);
        C2[5] = np.zeros(C2[5].size);
        C2[6] = np.zeros(C2[6].size);
        C2[7] = np.zeros(C2[7].size);
        baseline = pywt.waverec(C2,'sym8');
        x1 = denoiseWave - baseline;
        # plt.close('all')
        # plt.plot(x1)
        # plt.plot(resource)

    Wp=[0.9/500,50.0/500];
    Ws=[0.3/500,140.0/500];
    n, Wn = sci.buttord(Wp,Ws, 3, 10);
    b, a = sci.butter(n, Wn, 'bandpass');
    x2 = sci.filtfilt(b, a, x1);

    Wp1=[30.0/500,65.0/500];
    Ws1=[45.0/500,55.0/500];
    rp=3;
    rs=60;
    n1,Wn1 = sci.cheb2ord(Wp1,Ws1,rp,rs);
    b1,a1 = sci.cheby2(n1,rs,Wn1,'bandstop');
    re = sci.filtfilt(b1,a1,x2);

    return re;
